pkg/apiserver/registry/security/workloadidentity/storage/tokenrequest.go

```diff
@@ -0,0 +1,158 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/registry/rest"
+
+	"github.com/gardener/gardener/pkg/api"
+	securityapi "github.com/gardener/gardener/pkg/apis/security"
+	securityv1alpha1 "github.com/gardener/gardener/pkg/apis/security/v1alpha1"
+	securityvalidation "github.com/gardener/gardener/pkg/apis/security/validation"
+)
+
+// TokenRequestREST implements a RESTStorage for a token request.
+type TokenRequestREST struct {
+	// client kubernetes.Interface // Might be needed later to read the context object
+
+	minDuration            time.Duration
+	maxDuration            time.Duration
+	workloadIdentityGetter getter
+}
+
+type getter interface {
+	Get(context.Context, string, *metav1.GetOptions) (runtime.Object, error)
+}
+
+var (
+	_   = rest.NamedCreater(&TokenRequestREST{})
+	_   = rest.GroupVersionKindProvider(&TokenRequestREST{})
+	gvk = securityv1alpha1.SchemeGroupVersion.WithKind("TokenRequest")
+)
+
+// New returns an instance of the object.
+func (r *TokenRequestREST) New() runtime.Object {
+	return &securityv1alpha1.TokenRequest{}
+}
+
+// Destroy cleans up its resources on shutdown.
+func (r *TokenRequestREST) Destroy() {
+	// Given that underlying store is shared with REST, we don't destroy it here explicitly.
+}
+
+// Create returns a TokenRequest with workload identity token based on
+// - spec of the workload identity
+// - spec of the token request
+// - referenced context object
+// - gardener installation
+func (r *TokenRequestREST) Create(ctx context.Context, name string, obj runtime.Object, createValidation rest.ValidateObjectFunc, _ *metav1.CreateOptions) (runtime.Object, error) {
+	if createValidation != nil {
+		if err := createValidation(ctx, obj.DeepCopyObject()); err != nil {
+			return nil, err
+		}
+	}
+
+	tokenRequest := &securityapi.TokenRequest{}
+	if err := api.Scheme.Convert(obj, tokenRequest, nil); err != nil {
+		return nil, fmt.Errorf("failed converting %T to %T: %w", obj, tokenRequest, err)
+	}
+	if errs := securityvalidation.ValidateTokenRequest(tokenRequest); len(errs) != 0 {
+		return nil, apierrors.NewInvalid(gvk.GroupKind(), "", errs)
+	}
+
+	// TODO(vpnachev): implement context specific features
+	// if tokenRequest.Spec.ContextObject != nil {	}
+
+	wiObj, err := r.workloadIdentityGetter.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	wi, ok := wiObj.(*securityapi.WorkloadIdentity)
+	if !ok {
+		return nil, apierrors.NewInternalError(fmt.Errorf("cannot convert to *security.WorkloadIdentity object - got type %T", wiObj))
+	}
+
+	var (
+		aud = wi.Spec.Audiences
+		sub = wi.Status.Sub
+	)
+
+	duration := tokenRequest.Spec.Duration.Duration
+
+	if duration < r.minDuration {
+		duration = r.minDuration
+	} else if duration > r.maxDuration {
+		duration = r.maxDuration
+	}
```

Comment:
I think this will never execute since the validation will fail before that. See my other comment above